pkg/virt-launcher/virtwrap/live-migration-source.go

```diff
@@ -744,6 +782,139 @@ func generateMigrationParams(dom cli.VirDomain, vmi *v1.VirtualMachineInstance,
 	return params, nil
 }
 
+type getDiskVirtualSizeFuncType func(disk *libvirtxml.DomainDisk) (int64, error)
+
+var getDiskVirtualSizeFunc getDiskVirtualSizeFuncType
+
+func init() {
+	getDiskVirtualSizeFunc = getDiskVirtualSize
+}
+
+// getDiskVirtualSize return the size of a local volume to migrate.
+// See suggestion in: https://issues.redhat.com/browse/RHEL-4607
+func getDiskVirtualSize(disk *libvirtxml.DomainDisk) (int64, error) {
+	var path string
+	if disk.Source == nil {
+		return -1, fmt.Errorf("empty source for the disk")
+	}
+	switch {
+	case disk.Source.File != nil:
+		path = disk.Source.File.File
+	case disk.Source.Block != nil:
+		path = disk.Source.Block.Dev
+	default:
+		return -1, fmt.Errorf("not path set")
+	}
+	info, err := converter.GetImageInfo(path)
+	if err != nil {
+		return -1, err
+	}
+	return info.VirtualSize, nil
+}
+
+func getDiskName(disk *libvirtxml.DomainDisk) string {
+	if disk == nil {
+		return ""
+	}
+	n := disk.Alias.Name
+	if len(n) < 3 {
+		return n
+	}
+	// Trim the ua- prefix
+	if strings.HasPrefix(n, "ua-") {
+		return n[3:]
+	}
+	return n
+}
+
+func getMigrateVolumeForCondition(vmi *v1.VirtualMachineInstance, condition func(info *v1.StorageMigratedVolumeInfo) bool) map[string]bool {
+	res := make(map[string]bool)
+
+	for _, v := range vmi.Status.MigratedVolumes {
+		if v.SourcePVCInfo == nil || v.DestinationPVCInfo == nil {
+			continue
+		}
+		if v.SourcePVCInfo.VolumeMode == nil || v.DestinationPVCInfo.VolumeMode == nil {
+			continue
+		}
+		if condition(&v) {
+			res[v.VolumeName] = true
+		}
+	}
+	return res
+}
+
+func configureLocalDiskToMigrate(dom *libvirtxml.Domain, vmi *v1.VirtualMachineInstance) error {
+	if dom.Devices == nil {
+		return nil
+	}
+
+	migDisks := classifyVolumesForMigration(vmi)
+	fsSrcBlockDstVols := getMigrateVolumeForCondition(vmi, func(v *v1.StorageMigratedVolumeInfo) bool {
+		if v == nil {
+			return false
+		}
+		if *v.SourcePVCInfo.VolumeMode == k8sv1.PersistentVolumeFilesystem &&
+			*v.DestinationPVCInfo.VolumeMode == k8sv1.PersistentVolumeBlock {
+			return true
+		}
+		return false
+	})
+	blockSrcFsDstVols := getMigrateVolumeForCondition(vmi, func(v *v1.StorageMigratedVolumeInfo) bool {
+		if v == nil {
+			return false
+		}
+		if *v.SourcePVCInfo.VolumeMode == k8sv1.PersistentVolumeBlock &&
+			*v.DestinationPVCInfo.VolumeMode == k8sv1.PersistentVolumeFilesystem {
+			return true
+		}
+		return false
+	})
```

Comment:
I understand the desire to optimize and save space, but I think such on-the-fly function definitions harm readability. I would prefer to see a well defined function that explains what it does.